src/water_cmd.cpp

```diff
@@ -90,6 +90,17 @@ static void MarkCanalsAndRiversAroundDirty(TileIndex tile)
 	}
 }
 
+/**
+ * Clear non-flooding state of the tiles around a tile.
+ * @param tile The centre of the tile where other tiles' non-flooding state is cleared.
+ */
+void ClearNeighbourNonFloodingStates(TileIndex tile)
+{
+	for (Direction dir = DIR_BEGIN; dir != DIR_END; dir++) {
+		TileIndex dest = tile + TileOffsByDir(dir);
```

Comment:
```suggestion
		TileIndex dest = AddTileIndexDiffCWrap(tile, TileIndexDiffCByDir(dir));
```
Just preventing a situation where it would go to the other side of the map with `freeform_edges` turned off.